velox/functions/sparksql/tests/StringTest.cpp

```diff
@@ -497,5 +520,91 @@ TEST_F(StringTest, left) {
   EXPECT_EQ(left("da\u6570\u636Eta", 30), "da\u6570\u636Eta");
 }
 
+void StringTest::testTranslateConstantDictionary(
+    exec::ExprSet& exprSet,
+    const constant_dictionary_input_t& inputData) {
+  auto stringVector = makeFlatVector<StringView>(inputData.size());
+  for (int i = 0; i < inputData.size(); i++) {
+    stringVector->set(i, StringView(inputData[i].first));
+  }
+  auto result = evaluate(exprSet, makeRowVector({stringVector}));
+  FlatVector<StringView>* flatResult = result->asFlatVector<StringView>();
+  for (int32_t i = 0; i < inputData.size(); ++i) {
+    ASSERT_EQ(flatResult->valueAt(i), inputData[i].second);
+  }
+}
+
+void StringTest::testTranslateNonconstantDictionary(
+    exec::ExprSet& exprSet,
+    const nonconstant_dictionary_input_t& inputData) {
+  auto stringVector = makeFlatVector<StringView>(inputData.size());
+  auto matchVector = makeFlatVector<StringView>(inputData.size());
+  auto replaceVector = makeFlatVector<StringView>(inputData.size());
+  for (int i = 0; i < inputData.size(); i++) {
+    stringVector->set(i, StringView(std::get<0>(inputData[i].first)));
+    matchVector->set(i, StringView(std::get<1>(inputData[i].first)));
+    replaceVector->set(i, StringView(std::get<2>(inputData[i].first)));
+  }
+  auto result = evaluate(
+      exprSet, makeRowVector({stringVector, matchVector, replaceVector}));
+  FlatVector<StringView>* flatResult = result->asFlatVector<StringView>();
+  for (int32_t i = 0; i < inputData.size(); ++i) {
+    ASSERT_EQ(flatResult->valueAt(i), inputData[i].second);
+  }
+}
+
+TEST_F(StringTest, translate) {
+  EXPECT_EQ(translate("ab[cd]", "[]", "##"), "ab#cd#");
+  EXPECT_EQ(translate("ab[cd]", "[]", "#"), "ab#cd");
+  EXPECT_EQ(translate("ab[cd]", "[]", "#@$"), "ab#cd@");
+  EXPECT_EQ(translate("ab[cd]", "[]", "  "), "ab cd ");
+  EXPECT_EQ(translate("ab\u2028", "\u2028", "\u2029"), "ab\u2029");
+  EXPECT_EQ(translate("abcabc", "a", "\u2029"), "\u2029bc\u2029bc");
+  EXPECT_EQ(translate("abc", "", ""), "abc");
+  EXPECT_EQ(translate("translate", "rnlt", "123"), "1a2s3ae");
+  EXPECT_EQ(translate("translate", "rnlt", ""), "asae");
+  EXPECT_EQ(translate("abcd", "aba", "123"), "12cd");
+  // Test null input.
+  EXPECT_EQ(translate("abc", std::nullopt, "\u2029"), std::nullopt);
+  EXPECT_EQ(translate("abc", "\u2028", std::nullopt), std::nullopt);
+  EXPECT_EQ(translate(std::nullopt, "\u2028", "\u2029"), std::nullopt);
+
+  // Test constant dictionary cases.
```

Comment:
Thank you for iterating on the tests.

It is easier to debug test failures when tests are small. Hence, I suggest to break up this test into 3. I then suggest to remove helper methods and use lambdas. This will make tests shorter and all test code will be in one places making it easier to read. Here is an example:

```
TEST_F(StringTest, translateConstantMatch) {
  auto rowType = ROW({{"c0", VARCHAR()}});
  auto exprSet = compileExpression("translate(c0, 'ab', '12')", rowType);

  auto testTranslate = [&](const auto& input, const auto& expected) {
    auto result = evaluate(*exprSet, makeRowVector({input}));
    velox::test::assertEqualVectors(expected, result);
  };

  auto input = makeFlatVector<std::string>({"abcd", "cdab"});
  auto expected = makeFlatVector<std::string>({"12cd", "cd12"});
  testTranslate(input, expected);

  // Uses unicode batch as the next input.
  input = makeFlatVector<std::string>({"abåæçè", "åæçèab"});
  expected = makeFlatVector<std::string>({"12åæçè", "åæçè12"});
  testTranslate(input, expected);
}
```

Reply:
Yes, it makes sense. Just did some further code refactor according to your nice example. Thanks!